dbms/src/Core/SpillHandler.cpp

```diff
@@ -63,24 +76,37 @@ void SpillHandler::spillBlocks(const Blocks & blocks)
     try
     {
         Stopwatch watch;
-        RUNTIME_CHECK_MSG(spiller->spill_finished == false, "{}: spill after the spiller is finished.", spiller->config.spill_id);
+        RUNTIME_CHECK_MSG(spiller->isSpillFinished() == false, "{}: spill after the spiller is finished.", spiller->config.spill_id);
         auto block_size = blocks.size();
         LOG_INFO(spiller->logger, "Spilling {} blocks data into temporary file {}", block_size, current_spill_file_name);
         size_t total_rows = 0;
-        if (unlikely(writer == nullptr))
-        {
-            writer = std::make_unique<SpillWriter>(spiller->config.file_provider, current_spill_file_name, blocks[0].cloneEmpty(), spiller->spill_version);
-        }
+        size_t rows_in_file = 0;
+        size_t bytes_in_file = 0;
         for (const auto & block : blocks)
         {
-            total_rows += block.rows();
+            if (unlikely(writer == nullptr))
+            {
+                writer = std::make_unique<SpillWriter>(spiller->config.file_provider, current_spill_file_name, current_append_write, blocks[0].cloneEmpty(), spiller->spill_version);
+                rows_in_file = spilled_files[current_spilled_file_index]->getSpillDetails().rows;
+                bytes_in_file = spilled_files[current_spilled_file_index]->getSpillDetails().data_bytes_uncompressed;
+            }
+            auto rows = block.rows();
+            total_rows += rows;
+            rows_in_file += rows;
+            bytes_in_file += block.estimateBytesForSpill();
             writer->write(block);
+            if (spiller->enable_append_write && isSpilledFileFull(rows_in_file, bytes_in_file))
+            {
+                spilled_files[current_spilled_file_index]->updateSpillDetails(writer->finishWrite());
+                spilled_files[current_spilled_file_index]->markFull();
+                setUpNextSpilledFile();
+            }
         }
         double cost = watch.elapsedSeconds();
         time_cost += cost;
         LOG_INFO(spiller->logger, "Spilled {} rows from {} blocks into temporary file, time cost: {:.3f} sec.", total_rows, block_size, cost);
         RUNTIME_CHECK_MSG(current_spilled_file_index >= 0, "{}: spill after the spill handler is finished.", spiller->config.spill_id);
-        RUNTIME_CHECK_MSG(spiller->spill_finished == false, "{}: spill after the spiller is finished.", spiller->config.spill_id);
+        RUNTIME_CHECK_MSG(spiller->isSpillFinished() == false, "{}: spill after the spiller is finished.", spiller->config.spill_id);
```

Comment:
why check finished at the end of the work?

Reply:
Because spillBlocks not hold the lock, and it is not allowed to spill blocks after the spiller is finished, so add the check here to eary exit if error happens.